File: 2025/kotlin/src/main/kotlin/de/rubixdev/aoc/Day04.kt
package de.rubixdev.aoc

fun day4(input: String): Day = sequence {
    val example = """
    ..@@.@@@@.
    @@@.@.@.@@
    @@@@@.@.@@
    @.@@@@..@.
    @@.@@@@.@@
    .@@@@@@@.@
    .@.@.@.@@@
    @.@@@.@@@@
    .@@@@@@@@.
    @.@.@@@.@.
    """.trimIndent().trim()
//    val input = example.lines()
    val input = input.lines()
        .map { line -> line.map { it == '@' } }

    yield(null)
    yield(part1(input))
    yield(part2(input))
}

private fun part1(input: List<List<Boolean>>) = input.withIndex().sumOf { (y, line) ->
    line.withIndex().count { (x, tile) ->
        tile && ADJACENT.mapNotNull { input.getOrNull((x by y) + it) }.count { it } < 4
    }
}

private fun part2(input: List<List<Boolean>>) =
    generateSequence(0 to input) { (_, map) -> step(map) }
        .drop(1)
        .map { it.first }
        .takeWhile { it != 0 }
        .sum()

private fun step(map: List<List<Boolean>>) = map.withIndex().map { (y, line) ->
    line.withIndex().map { (x, tile) ->
        when (tile && ADJACENT.mapNotNull { map.getOrNull((x by y) + it) }.count { it } < 4) {
            true -> 1 to false
            false -> 0 to tile
        }
    }.foldFirst(0) { sum, count -> sum + count }
}.foldFirst(0) { sum, count -> sum + count }

File: 2025/scala/src/main/scala/de/rubixdev/aoc/Day04.scala
package de.rubixdev.aoc

object Day04 extends Day {
  override def run(rawInput: String): Unit = {
    val input = rawInput.linesIterator
      .map(_.map(_ == '@'))
      .toList

    println("Part 1: " + part1(input))
    println("Part 2: " + part2(input))
  }

  private def shouldRemove(
      map: Seq[Seq[Boolean]],
      tile: Boolean,
      x: Int,
      y: Int,
  ): Boolean =
    tile && ADJACENT
      .flatMap { offset => map.lift2d((x by y) + offset) }
      .count(identity) < 4

  private def part1(input: List[Seq[Boolean]]) = input.view.zipWithIndex.map {
    (line, y) =>
      line.zipWithIndex.count { (tile, x) => shouldRemove(input, tile, x, y) }
  }.sum

  private def part2(input: List[Seq[Boolean]]) =
    LazyList
      .iterate(0 -> (input: Seq[Seq[Boolean]])) { (_, map) => step(map) }
      .drop(1)
      .map(_._1)
      .takeWhile(_ != 0)
      .sum

  private def step(map: Seq[Seq[Boolean]]) = map.view.zipWithIndex
    .map { (line, y) =>
      line.view.zipWithIndex
        .map { (tile, x) =>
          if (shouldRemove(map, tile, x, y)) 1 -> false
          else 0 -> tile
        }
        .foldLeftFirst(0)(_ + _)
    }
    .foldLeftFirst(0)(_ + _)
}
